File: leetcode/84.cpp
class Solution {
public:
  int largestRectangleArea(vector<int>& heights) {
    int n = heights.size();
    int maxArea = 0;
    vector<pair<int,int>> vstack;
    for (int i = 0; i < n; i++) {
      int h = heights[i];
      int p = i;
      while (!vstack.empty() && vstack.back().first > h) {
        pair<int,int> x = vstack.back();
        p = x.second;
        maxArea = max(maxArea, max(h*((i-p)+1), x.first*(i-p)));
        vstack.pop_back();
      }
      vstack.push_back({h,p});
    }
    for(auto t: vstack) maxArea = max(maxArea, t.first*(n-t.second));
    return maxArea;
  }
}; 

File: leetcode/84.py
from typing import List

class Solution:
  def largestRectangleArea(self, heights:List[int]) -> int:
    maxArea = 0
    stack = []
    for i, h in enumerate(heights):
      p = i
      while stack and stack[-1][0] > h:
        g, j = stack.pop()
        p = j
        maxArea = max(maxArea, h*(i-j+1), g*(i-j))
      stack.append((h,p))
    for h, i in stack: maxArea = max(maxArea, h*(len(heights)-i))
    return maxArea

